merkle/src/tree_impl.rs

```diff
@@ -0,0 +1,182 @@
+            leafs: Vec::with_capacity(capacity),
+            nodes: Vec::with_capacity(protential_node_count),
+            ..Default::default()
+        }
+    }
+
+    fn clear_dirty_hashes(&mut self, leaf_index: usize) {
+        let mut parent = leaf_index;
+        while parent > 0 {
+            parent = calculate_parent_index(parent);
+            if self.nodes[parent].is_some() {
+                self.nodes[parent] = None;
+            } else {
+                break;
+            }
+        }
+    }
+
+    fn calculate_hash_if_needed(
+        &mut self,
+        index: usize,
+    ) -> Option<(THash, MerkleTreeNodeMetadata)> {
+        if index < self.nodes.len() {
+            if let Some(hash) = &self.nodes[index] {
+                Some((hash.clone(), MerkleTreeNodeMetadata::new(index)))
+            } else {
+                let left = index * 2 + 1;
+                let right = index * 2 + 2;
+                let left_hash = self.calculate_hash_if_needed(left);
+                let right_hash = self.calculate_hash_if_needed(right);
+                let hash = TMerger::merge(&left_hash, &right_hash);
+                self.nodes[index] = hash.clone();
+                hash.map(|hash| (hash, MerkleTreeNodeMetadata::new(index)))
+            }
+        } else {
+            let leaf_index = index - self.nodes.len();
+            if leaf_index < self.leafs.len() {
+                let (hash, _) = &self.leafs[leaf_index];
+                Some((hash.clone(), MerkleTreeNodeMetadata::new(index)))
+            } else {
+                None
+            }
+        }
+    }
+}
+
+impl<TItem, THash, THasher, TMerger> MerkleTree for MinaMerkleTree<TItem, THash, THasher, TMerger>
+where
+    THasher: MerkleHasher<Item = TItem, Hash = THash>,
+    TMerger: MerkleMerger<Hash = THash>,
+    THash: Clone,
+{
+    type Item = TItem;
+    type Hash = THash;
+
+    fn depth(&self) -> u32 {
+        self.depth
+    }
+
+    fn count(&self) -> usize {
+        self.leafs.len()
+    }
+
+    fn root(&mut self) -> Option<Self::Hash> {
+        self.calculate_hash_if_needed(0).map(|(hash, _)| hash)
+    }
+
+    fn add_batch(&mut self, items: Vec<Self::Item>) {
+        let new_leaf_count = self.leafs.len() + items.len();
+        let new_depth = calculate_depth(new_leaf_count);
+        if new_depth != self.depth {
+            let new_node_count = calculate_node_count(new_depth);
+            self.depth = new_depth;
+            self.nodes = vec![None; new_node_count];
+        } else {
+            let start = self.nodes.len() + self.leafs.len();
+            for i in start..(start + items.len()) {
+                self.clear_dirty_hashes(i);
+            }
+        }
+        for leaf in items
+            .into_iter()
+            .enumerate()
+            .map(|(i, item)| (THasher::hash(&item, MerkleTreeNodeMetadata::new(i)), item))
+        {
+            self.leafs.push(leaf);
+        }
+    }
+
+    fn add(&mut self, item: Self::Item) {
```

Comment:
is add supposed to create a new `vec![]` everytime for adding the passed `item`

Reply:
No it creates a new `vec![]` only when tree depth / height is changed, in that case all hashes need to be recalculated anyway